### PR TITLE
solid_dmft parser 

### PR DESCRIPTION
**Pull request by JosePizarro3**
And changes to adapt new metainfo in w2dynamics.

### CODE REVIEW

**Comment by JosePizarro3** [comment]
Hi @ladinesa ,

I was working on a new DMFT parser. This is linked with a new branch in nomad (https://gitlab.mpcdf.mpg.de/nomad-lab/nomad-FAIR/-/commits/1264-add-solid_dmft-parser). Some quantities (specially for Greens Functions) have been slightly modified to address multi-site calculations (i.e. now they have an extra dimension).

I thus had to modify the w2dynamics parser too to adapt to this changes.

Could you please review it? No rush, as I am still waiting for the first DMFT branch (1156) to be merge in nomad.
Also, I am not fully happy with the 'parse_scc' method. Maybe you can take a closer look and give me suggestions on how to optimize it :slightly_smiling_face:

File: electronicparsers/soliddmft/parser.py [hunk 1]
```diff
@@ -0,0 +1,399 @@
...
+        for keys in self._solver_map.keys():
+            if sec_method.x_soliddmft_general.x_soliddmft_solver_type == keys:
+                sec_dmft.impurity_solver = self._solver_map[keys]
+                break
+
+    def parse_scc(self, data):
+        sec_run = self.archive.run[-1]
+        sec_scc = sec_run.m_create(Calculation)
+        if hasattr(self.archive.run[-1], 'system') and len(self.archive.run[-1].system) > 0:
+            sec_scc.system_ref = sec_run.system[-1]
```
Comment (ladinesa) [inline electronicparsers/soliddmft/parser.py]: `sec_scc.system_ref = sec_run.system[-1] if sec_run.system else None`

File: electronicparsers/soliddmft/parser.py [hunk 2]
```diff
@@ -0,0 +1,399 @@
...
+                        if s == 'block_names':
+                            continue
+                        param.append(self.dmft_results[it_key][gf_key_mod][s]['data'][()])
+                param = np.array(param)
+                setattr(sec_iter, f'x_soliddmft_{gf_key}', param[:, :, 0, 0, :])
+
+        # SCF steps
+        scf_keys = [int(key.lstrip('it_')) for key in self.dmft_results.keys() if key.startswith('it_')]
+        scf_keys.sort()
+        scf_keys_sorted = ['it_' + str(key) for key in scf_keys]
```
Comment (ladinesa) [inline electronicparsers/soliddmft/parser.py]: use f'' formatting

File: electronicparsers/soliddmft/parser.py [hunk 3]
```diff
@@ -0,0 +1,399 @@
...
+        self.filepath = filepath
+        self.archive = archive
+        self.maindir = os.path.dirname(self.filepath)
+        self.logger = logging.getLogger(__name__) if logger is None else logger
+
+        try:
+            data = h5py.File(self.filepath)
+        except Exception:
+            self.logger.error('Error opening h5 file.')
+            data = None
```
Comment (ladinesa) [inline electronicparsers/soliddmft/parser.py:360]: return

File: electronicparsers/soliddmft/parser.py [hunk 4]
```diff
@@ -0,0 +1,399 @@
...
+
+        if data is None:
```
Comment (ladinesa) [inline electronicparsers/soliddmft/parser.py]: remove

File: electronicparsers/soliddmft/parser.py [hunk 5]
```diff
@@ -0,0 +1,399 @@
...
+        scf_keys_sorted = ['it_' + str(key) for key in scf_keys]
+
+        # We store the last post-processed step for observables
+        total_iterations = self.dmft_results.get('iteration_count', 1)[()]
+        for it in range(total_iterations + 1):
+            sec_scf_iteration = sec_scc.m_create(ScfIteration)
+
+            if it < total_iterations:
+                # iterations quantities
+                it_key = 'it_' + str(it + 1)
```
Comment (ladinesa) [inline electronicparsers/soliddmft/parser.py]: use f'' formatting

File: electronicparsers/soliddmft/parser.py [hunk 6]
```diff
@@ -0,0 +1,399 @@
...
+    def parse_scc(self, data):
+        sec_run = self.archive.run[-1]
+        sec_scc = sec_run.m_create(Calculation)
+        if hasattr(self.archive.run[-1], 'system') and len(self.archive.run[-1].system) > 0:
+            sec_scc.system_ref = sec_run.system[-1]
+        sec_scc.method_ref = sec_run.method[-1]  # ref to DMFT
+
+        def parse_iteration_quantities(scf_section, it_key):
+            sec_iter = scf_section.m_create(x_soliddmft_iter_parameters)
+            for en_key in self.iteration_energies:
```
Comment (ladinesa) [inline electronicparsers/soliddmft/parser.py]: personally, given that there are only a few keys,i woud not use setattr rather set the quantities directly, after all they have differ in the shape and an extra reshaping in some other quantities.

Comment (ladinesa) [reply electronicparsers/soliddmft/parser.py]: If you want to keep this loop over all keys, I suggest that you do a recursion to extract the datasets over nested groups.

File: electronicparsers/soliddmft/parser.py [hunk 7]
```diff
@@ -0,0 +1,399 @@
...
+                    for s in self.dmft_results[it_key][gf_key_mod].keys():
+                        if s == 'block_names':
+                            continue
+                        param.append(self.dmft_results[it_key][gf_key_mod][s]['data'][()])
+                param = np.array(param)
+                setattr(sec_iter, f'x_soliddmft_{gf_key}', param[:, :, 0, 0, :])
+
+        # SCF steps
+        scf_keys = [int(key.lstrip('it_')) for key in self.dmft_results.keys() if key.startswith('it_')]
+        scf_keys.sort()
```
Comment (ladinesa) [inline electronicparsers/soliddmft/parser.py:262]: sort directly without stripping it_

File: electronicparsers/soliddmft/parser.py [hunk 8]
```diff
@@ -0,0 +1,399 @@
...
+                    param = self.dmft_results['observables']['mu'][str(it)]
+                else:
+                    param = []
+                    for i in self.dmft_results['observables'][keys].keys():
+                        for s in self.dmft_results['observables'][keys][i].keys():
+                            param.append(self.dmft_results['observables'][keys][i][s][str(it)][()])
+                setattr(sec_obs, f'x_soliddmft_{keys}', param)
+
+            # Chemical potential
+            if hasattr(sec_scc.scf_iteration[it], 'x_soliddmft_observables') and sec_scc.scf_iteration[it].x_soliddmft_observables.x_soliddmft_mu is not None:
```
Comment (ladinesa) [inline electronicparsers/soliddmft/parser.py]: doesn't `if sec_scc.scf_iteration[it].x_soliddmft_observables` work?

File: electronicparsers/soliddmft/parser.py [hunk 9]
```diff
@@ -0,0 +1,399 @@
...
+        beta = sec_scc.method_ref.x_soliddmft_general.x_soliddmft_beta
+        n_tau = sec_scc.method_ref.x_soliddmft_general.x_soliddmft_n_tau
+        n_iw = sec_scc.method_ref.x_soliddmft_general.x_soliddmft_n_iw
+        sec_gf.tau = [n * beta / (n_tau - 1) for n in range(n_tau)]
+        sec_gf.matsubara_freq = [(2 * (n - n_iw) + 1) / beta for n in range(2 * n_iw)]
+        sec_gf.chemical_potential = sec_scc.scf_iteration[-1].x_soliddmft_observables.x_soliddmft_mu
+        nat = sec_scc.method_ref.dmft.n_atoms_per_unit_cell
+        norb = sec_scc.method_ref.dmft.n_correlated_orbitals
+        for keys in self._gf_map.keys():
+            funct = self.dataset_to_complex(getattr(sec_scc.scf_iteration[-1].x_soliddmft_iter, f'x_soliddmft_{keys}'))
```
Comment (ladinesa) [inline electronicparsers/soliddmft/parser.py]: no need to define function here as used only once

File: electronicparsers/soliddmft/parser.py [hunk 10]
```diff
@@ -0,0 +1,399 @@
...
+                            param.append(self.dmft_results[it_key][en_key][i][()])
+                    param = np.array(param)
+                    if en_key.startswith('DC'):
+                        setattr(sec_iter, f'x_soliddmft_{en_key}', param)
+                    else:  # deleting useless indices
+                        setattr(sec_iter, f'x_soliddmft_{en_key}', param[:, 0, 0, :])
+            for gf_key in self.iteration_gfs:
+                param = []
+                for i in range(sec_scc.method_ref.dmft.n_atoms_per_unit_cell):  # atom index
+                    gf_key_mod = gf_key + '_' + str(i)
```
Comment (ladinesa) [inline electronicparsers/soliddmft/parser.py]: use f'' formatting

File: electronicparsers/soliddmft/parser.py [hunk 11]
```diff
@@ -0,0 +1,399 @@
...
+            'hubbardI': 'hubbard_I',
+            'ctseg': 'CT-HYB'
+        }
+
+        self.iteration_energies = [
+            'DC_energ', 'DC_pot', 'dens_mat_pre', 'dens_mat_post', 'chemical_potential_pre',
+            'chemical_potential_post'
+        ]
+
+        self.iteration_gfs = ['Delta_time', 'G0_freq', 'Gimp_freq', 'Gimp_time', 'Sigma_freq']
```
Comment (ladinesa) [inline electronicparsers/soliddmft/parser.py:70]: can's this list and mapping go in one dict?

Comment (JosePizarro3) [reply electronicparsers/soliddmft/parser.py:70]: which ones you mean, `self.iteration_gfs` and `self._gf_map`? If so, `_gf_map` does not contain all `iteration_gfs` quantities, so is there a way of defining a dict there?

File: electronicparsers/soliddmft/parser.py (hunk 1 above)
Comment (JosePizarro3) [reply electronicparsers/soliddmft/parser.py]: done

File: electronicparsers/soliddmft/parser.py (hunk 2 above)
Comment (JosePizarro3) [reply electronicparsers/soliddmft/parser.py]: done

File: electronicparsers/soliddmft/parser.py (hunk 3 above)
Comment (JosePizarro3) [reply electronicparsers/soliddmft/parser.py:360]: done

File: electronicparsers/soliddmft/parser.py (hunk 4 above)
Comment (JosePizarro3) [reply electronicparsers/soliddmft/parser.py]: done

File: electronicparsers/soliddmft/parser.py (hunk 5 above)
Comment (JosePizarro3) [reply electronicparsers/soliddmft/parser.py]: done

File: electronicparsers/soliddmft/parser.py (hunk 6 above)
Comment (JosePizarro3) [reply electronicparsers/soliddmft/parser.py]: check out whether I got your suggestion

we can also chat quickly over zoom if I didn't

File: electronicparsers/soliddmft/parser.py (hunk 8 above)
Comment (JosePizarro3) [reply electronicparsers/soliddmft/parser.py]: it does :slightly_smiling_face: 

I will stop using `hasattr` so often, my bad

File: electronicparsers/soliddmft/parser.py (hunk 9 above)
Comment (JosePizarro3) [reply electronicparsers/soliddmft/parser.py]: function deleted, and operation moved directly to line

File: electronicparsers/soliddmft/parser.py (hunk 10 above)
Comment (JosePizarro3) [reply electronicparsers/soliddmft/parser.py]: done

File: electronicparsers/soliddmft/parser.py (hunk 11 above)
Comment (ladinesa) [reply electronicparsers/soliddmft/parser.py:70]: ah ok when they are different then it is ok.

**Comment by JosePizarro3** [comment]
Done with your first round of comments. Feel free to resolve them or further continue the conversation.

Thanks a lot! :v:

**Comment by ladinesa** [comment]
great. please merge.

**Pull request merged**